misc: Update supported browser messages

0v00 opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes [https://github.com/cypress-io/cypress/issues/28914](https://github.com/cypress-io/cypress/issues/28914)

### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->
Current messages for browser support/browser's not found can be confusing e.g.:
`Browser: chrome was not found on your system or is not supported by Cypress.`
This PR aims to provide clear messaging if a browser is not supported, or if it was not found by name or path.
- If a user enters a browser name or path that does not contain any supported browser names, or does not contain any custom browser (e.g. Brave) names added via Cypress config, then the user will see a `BROWSER_NOT_SUPPORTED` error message.
- I have added a regex that should cast a wide net for supported browsers - we ideally want to have zero false negatives. If a user enters a browser name or path that is supported (e.g. `C:\some\path\msedge.exe`) but is not found on their system, they will still see the appropriate `BROWSER_NOT_FOUND_BY_NAME` or `BROWSER_NOT_FOUND_BY_PATH` error. However, if a user misspells a browser name e.g. `webkitt`, or enters an unsupported browser name e.g. `arc`, they will see the `BROWSER_NOT_SUPPORTED` error message.
- I have moved the special `canary` error message to the `BROWSER_NOT_SUPPORTED` error message.
- `getBrowserLauncher()` now throws the `BROWSER_NOT_SUPPORTED` error instead of `BROWSER_NOT_FOUND_BY_NAME` - if `electron` or one of the supported browser families is not found, then it should be safe to assume that this browser is not supported.
- I have added `webkit` to the list of supported browsers in the `BROWSER_NOT_SUPPORTED` error message.
- Custom browsers added via Cypress config still work and are not filtered out.
- I have updated the `BROWSER_NOT_FOUND_BY_NAME` error - this error message may still need to be made clearer.
- This PR doesn't take channels into account e.g. passing in `chrome:someunsupportedchannel` - this will throw a `BROWSER_NOT_FOUND_BY_NAME` error. I think this behavior makes sense as we want to avoid any false negatives - if the name/path contains e.g. `chrome`, we can assume it's supported and then throw a more appropriate error later on.
- _*(Apologies for the number of commits)*_

### Steps to test
<!--
For non-trivial behavior changes, list the steps that a reviewer should follow to validate the new behavior.
This is not meant to be the only testing performed by a reviewer, just the "happy path" that leads to the new behavior.
-->
Test the following commands:
- `yarn cypress:run --browser unsupportedbrowser --project /your/project`
- `yarn cypress:run --browser badname:channel--project /your/project`
- `yarn cypress:run --browser /path/to/unsupported/browser --project /your/project`
- Run `yarn test test/unit/browsers/browsers_spec.js` to verify passing tests
- Run `yarn test` in `packages/errors` to verify passing tests

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->
- If a user enters a browser name or path that does not match any supported browser names, or does not match any custom browsers (e.g. Brave) added in via Cypress config, then the user will see a 'browser not supported' error message.
<img width="608" alt="unsupported_path" src="https://github.com/cypress-io/cypress/assets/154560623/4f710167-b780-444d-b649-18f3ff284263">
<img width="610" alt="unsupported_name" src="https://github.com/cypress-io/cypress/assets/154560623/94979c3a-cd60-4f37-b166-aeea80911602">
<img width="606" alt="unsupported_name_channel" src="https://github.com/cypress-io/cypress/assets/154560623/7540ea07-8108-4c2d-8bdc-8cb93e20d9d6">


### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

0v00 committed:
add browser not supported error

0v00 committed:
add browser not supported error

0v00 committed:
export knownBrowsers

0v00 committed:
export knownBrowsers

0v00 committed:
update browser_not_found_by_name err, add webkit to supported browser message

0v00 committed:
update browser_not_found_by_name err, add webkit to supported browser message

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=28964) <br/>All committers have signed the CLA.

0v00 committed:
add to changelog

0v00 committed:
add to changelog

jennifer-shehane commented:
@0v00 Could you sign our CLA? And also add a test around this?

0v00 committed:
remove console.logs

0v00 committed:
remove console.logs

0v00 committed:
update comment

0v00 committed:
update comment

0v00 committed:
add return type

0v00 committed:
add return type

0v00 committed:
update browser cannot be found spec

0v00 committed:
update browser cannot be found spec

0v00 committed:
update expect exit error

0v00 committed:
update expect exit error

0v00 committed:
update rejectedwith type

0v00 committed:
update rejectedwith type

0v00 committed:
update browsers_spec tests and snapshots

0v00 committed:
update browsers_spec tests and snapshots

0v00 committed:
add fn for throwing when browser not supported

0v00 committed:
add fn for throwing when browser not supported

0v00 committed:
add browser_not_supported error'

0v00 committed:
add browser_not_supported error'

0v00 committed:
change to browser

0v00 committed:
change to browser

0v00 committed:
update browser support logic and comments

0v00 committed:
update browser support logic and comments

0v00 committed:
update tests and snapshots

0v00 committed:
update tests and snapshots

0v00 committed:
remove changelog entry

0v00 committed:
remove changelog entry

0v00 committed:
update errors and snapshots

0v00 committed:
update errors and snapshots

0v00 committed:
update comment

0v00 committed:
update comment

0v00 committed:
update comments, update tests, change from not found to not supported err

0v00 committed:
update comments, update tests, change from not found to not supported err

0v00 committed:
update changelog

0v00 committed:
update changelog

0v00 commented:
> @0v00 Could you sign our CLA? And also add a test around this?

Added and updated tests, signed the CLA and updated the changelog. Marked as ready for review. Apologies for the number of commits.

0v00 committed:
fix wording in test case, update snapshots

0v00 committed:
fix wording in test case, update snapshots

0v00 committed:
remove not_found_by_path test as may be hard to test properly in ci

0v00 committed:
remove not_found_by_path test as may be hard to test properly in ci

jennifer-shehane committed:
Merge branch 'develop' into issue-28914-browser-support-message

jennifer-shehane commented on packages/server/lib/browsers/utils.ts:
```diff
@@ -338,12 +338,44 @@ const parseBrowserOption = (opt) => {
+ * Cast a wide net when checking for browser support - this should avoid all false negatives
+ * The regex below will match: msedge.exe, edge, edge-beta
+ */
+function createSupportedBrowserRegex (browsers: string[]): RegExp {
```
I'm remembering that @emilyrohrbough had a draft PR to fix the supported browsers list because it's currently not accurate in what it says we support. https://github.com/cypress-io/cypress/pull/28377/files Maybe she can comment or this code will be helpful somehow.

0v00 replied:
That was very helpful thank you - a few thoughts around adding `webkit` and `electron` directly to the `knownBrowsers` object:
- Does the Electron browser have a binary or a versionRegex? If not, I wonder if it might be better to export it separately in `known-browsers` with its own type defined in `types/src/browsers.ts` e.g. `export type ElectronBrowser = Omit<Browser, 'versionRegex' | 'binary'>`. Otherwise we could make the `binary` and `versionRegex` properties in `Browser` optional and add Electron to `knownBrowsers`, this would require a few changes in `launcher/lib/detect.ts` (not sure what else we would have to change).

- The Webkit browser in @emilyrohrbough draft PR conforms to the `Browser` type, but adding it directly to the `knownBrowsers` object will require some changes to `launcher/lib/windows/index.ts` and `launcher/lib/darwin/index.ts` in order to get the `darwin_spec.ts` and `windows_spec.ts` tests passing. Same for Electron.

- Adding these both directly to the `knownBrowsers` I believe will require some additional changes to browser detection in `server/lib/browsers/utils.ts`, specifically in `getBrowsers()`

I think the potential options are: 
- add both `webkit` and `electron` to `known-browsers.ts` but exported separately from the `knownBrowsers` object (and look at properly integrating webkit and electron into knownBrowsers in another PR)
- add them into the `knownBrowsers` object in this PR (but this PR will get a bit bigger in scope)

I'm not sure if there is a better solution that I'm missing entirely (I could just be overthinking this).

github-actions[bot] committed:
chore: Update Chrome (beta) to 123.0.6312.22 (#29041)

Co-authored-by: cypress-bot[bot] <41898282+cypress-bot[bot]@users.noreply.github.com>

github-actions[bot] committed:
chore: Update v8 snapshot cache (#29039)

* chore: updating v8 snapshot cache

* chore: updating v8 snapshot cache

* chore: updating v8 snapshot cache

---------

Co-authored-by: cypress-bot[bot] <+cypress-bot[bot]@users.noreply.github.com>
Co-authored-by: Jennifer Shehane <jennifer@cypress.io>

renovate[bot] committed:
dependency: update dependency signal-exit to v3.0.7 (#28979)

* fix(deps): update dependency signal-exit to v3.0.7

* Remove signal-exit - this is not used

* empty commit

* fix(deps): update dependency signal-exit to v3.0.7

* Update changelog

* Update cli/CHANGELOG.md

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Jennifer Shehane <jennifer@cypress.io>
Co-authored-by: Emily Rohrbough <emilyrohrbough@users.noreply.github.com>
Co-authored-by: Ryan Manuel <ryanm@cypress.io>

0v00 committed:
update knownBrowsers, tests, types

0v00 committed:
remove import
